### PR TITLE
Fix "RuntimeError: Example is wrong shape"

### LINKED ISSUE #410: NWPDataSource RuntimeError: Example is wrong shape!
```
2021-11-16 18:54:25,868 DEBUG processID=1365921 Opening NWP data: /mnt/storage_ssd_4tb/data/ocf/solar_pv_nowcasting/nowcasting_dataset_pipeline/NWP/UK_Met_Office/UKV/zarr/UKV_intermediate_version_2.zarr | /home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/nwp/nwp_data_source.py#L169
2021-11-16 18:54:28,338 WARNING processID=1365921 NWP Zarr has 6 duplicated init_times.  Fixing... | /home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/nwp/nwp_data_source.py#L182
2021-11-16 18:54:28,340 WARNING processID=1365921 NWP Zarr init_time is not monotonic_increasing.  Fixing... | /home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/nwp/nwp_data_source.py#L189
2021-11-16 18:54:28,352 INFO processID=1365921 Fixed 12 out of order init_times. | /home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/nwp/nwp_data_source.py#L197
2021-11-16 18:54:28,359 DEBUG processID=1365921 NWPDataSource creating batch 1! | /home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/data_source.py#L199
2021-11-16 18:54:28,688 ERROR processID=1365509 Worker process nwp raised exception! | /home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/manager.py#L449
NoneType: None
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/concurrent/futures/process.py", line 243, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/data_source.py", line 202, in create_batches
    batch = self.get_batch(
  File "/home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/data_source.py", line 260, in get_batch
    examples = [future_example.result() for future_example in future_examples]
  File "/home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/data_source.py", line 260, in <listcomp>
    examples = [future_example.result() for future_example in future_examples]
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/concurrent/futures/_base.py", line 438, in result
    return self.__get_result()
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/data_sources/data_source.py", line 419, in get_example
    raise RuntimeError(
RuntimeError: Example is wrong shape! x_meters_center=241537.6109145787
y_meters_center=639792.3593665359
t0_dt=2020-10-09 12:30:00
times are <xarray.DataArray 'time' (time: 3)>
array(['2020-10-09T12:00:00.000000000', '2020-10-09T13:00:00.000000000',
       '2020-10-09T14:00:00.000000000'], dtype='datetime64[ns]')
Coordinates:
    init_time  datetime64[ns] 2020-10-09T09:00:00
    step       (time) timedelta64[ns] 03:00:00 04:00:00 05:00:00
  * time       (time) datetime64[ns] 2020-10-09T12:00:00 ... 2020-10-09T14:00:00
Attributes:
    long_name:      time since forecast_reference_time
    standard_name:  forecast_period
expected shape=(10, 4, 64, 64)
actual shape (10, 3, 0, 64)
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jack/dev/ocf/nowcasting_dataset/scripts/prepare_ml_data.py", line 79, in <module>
    main()
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/utils.py", line 155, in inner_func
    return func(*args, **kwargs)
  File "/home/jack/dev/ocf/nowcasting_dataset/scripts/prepare_ml_data.py", line 72, in main
    manager.create_batches(overwrite_batches)
  File "/home/jack/dev/ocf/nowcasting_dataset/nowcasting_dataset/manager.py", line 450, in create_batches
    raise exception
RuntimeError: Example is wrong shape! x_meters_center=241537.6109145787
y_meters_center=639792.3593665359
t0_dt=2020-10-09 12:30:00
times are <xarray.DataArray 'time' (time: 3)>
array(['2020-10-09T12:00:00.000000000', '2020-10-09T13:00:00.000000000',
       '2020-10-09T14:00:00.000000000'], dtype='datetime64[ns]')
Coordinates:
    init_time  datetime64[ns] 2020-10-09T09:00:00
    step       (time) timedelta64[ns] 03:00:00 04:00:00 05:00:00
  * time       (time) datetime64[ns] 2020-10-09T12:00:00 ... 2020-10-09T14:00:00
Attributes:
    long_name:      time since forecast_reference_time
    standard_name:  forecast_period
expected shape=(10, 4, 64, 64)
actual shape (10, 3, 0, 64)
```

### PR DESCRIPTION
**Pull request by JackKelly**
# Pull Request

*PLEASE NOTE THAT THERE ARE ONLY TWO SMALL CHANGES IN THIS PR IN `nwp_data_source.py`: [HERE](https://github.com/openclimatefix/nowcasting_dataset/pull/412/files#diff-419324154fec1cae483a63ac8349dd8c2e2b0540fa9022f4f424de830b80d671R26) AND [HERE](https://github.com/openclimatefix/nowcasting_dataset/pull/412/files#diff-419324154fec1cae483a63ac8349dd8c2e2b0540fa9022f4f424de830b80d671R184)*

The reason GitHub says 7 files have changed is because I merged a bunch of other bug fixes into this branch. Specifically:

* jack/log-to-files (PR #404)
* bug/manger-sun (PR #403)
* bug/400-pandas-loc

Good news:  `prepare_ml_data.py` now creates NWP batches!  And fairly fast, too (about 1 per second).

Fixes #410

## How Has This Been Tested?

- [ ] No
- [x] Yes, unit tests pass
- [x] `prepare_ml_data.py` runs successfully

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: nowcasting_dataset/data_sources/nwp/nwp_data_source.py [hunk 1]
```diff
@@ -181,6 +181,16 @@ def open_nwp(zarr_path: str, consolidated: bool) -> xr.DataArray:
 
     ukv = nwp["UKV"]
```
Comment (peterdudfield) [inline nowcasting_dataset/data_sources/nwp/nwp_data_source.py:182]: is the reversing needed for other channels?

Comment (peterdudfield) [reply nowcasting_dataset/data_sources/nwp/nwp_data_source.py:182]: I may have miss understood here - UKV might not be a channel. Perhaps add a comment why "UKV" is selected

**Review by peterdudfield — APPROVED** [review]
Looks good. 

- just a small comment
- looks like a few conflicts, but im sure you can merge main into this first

File: nowcasting_dataset/data_sources/nwp/nwp_data_source.py (hunk 1 above)
Comment (JackKelly) [reply nowcasting_dataset/data_sources/nwp/nwp_data_source.py:182]: Ah, yes, good point!  `UKV` is the name of the (one and only) `DataArray` in the Zarr `Dataset`.  You're right, `UKV` isn't a channel.  `UKV` is the name of the DataArray which holds all the channels (the channels are represented as a dimension in the DataArray).  I'll add a comment!  Thanks!

**Pull request merged**